internal/server/data/migrations.go

```diff
@@ -167,7 +166,25 @@ func addAuthURLAndScopeToProviders() *migrator.Migration {
 // #2360: delete duplicate grants (the same subject, resource, and privilege) to allow for unique constraint
 func deleteDuplicateGrants() *migrator.Migration {
 	return &migrator.Migration{ID: "202207081217", Migrate: func(tx *gorm.DB) error {
-		return deleteDuplicates(tx, "subject, resource, privilege", &models.Grant{})
+		stmt := `
+			DELETE FROM grants
+			WHERE deleted_at IS NULL
+			AND id NOT IN (
+				SELECT min(id)
+				FROM grants
+				WHERE deleted_at IS NULL
+				GROUP BY (subject, resource, privilege))`
+		if err := tx.Exec(stmt).Error; err != nil {
+			return err
+		}
+
+		stmt = `CREATE UNIQUE INDEX idx_grant_srp ON grants
+USING btree (subject, privilege, resource)
+WHERE (deleted_at IS NULL);`
+		if err := tx.Exec(stmt).Error; err != nil {
+			return err
+		}
```

Comment:
I added the index explicitly here in preparation for removing `gorm.AutoMigrate`. This statement was copied from a DB dump, so it should match exactly what `gorm.AutoMigrate` produces.